zerver/lib/url_decoding.py

```diff
@@ -24,3 +28,127 @@ def is_same_server_message_link(url: str) -> bool:
     ends_with_near_message_id = fragment_parts[-2] == "near" and fragment_parts[-1].isdigit()
 
     return category == "narrow" and section in {"channel", "dm"} and ends_with_near_message_id
+
+
+CHANNEL_SYNONYMS = {"stream": "channel", "streams": "channels"}
+
+OPERATOR_SYNONYMS = {
+    **CHANNEL_SYNONYMS,
+    # "pm-with:" was renamed to "dm:"
+    "pm-with": "dm",
+    # "group-pm-with:" was replaced with "dm-including:"
+    "group-pm-with": "dm-including",
+    "from": "sender",
+    DB_TOPIC_NAME: "topic",
+}
+
+
+def canonicalize_operator_synonyms(text: str) -> str:
+    text = text.lower()
+    if text in OPERATOR_SYNONYMS.values():
+        return text
+    if text in OPERATOR_SYNONYMS:
+        return OPERATOR_SYNONYMS[text]
+    return text
+
+
+def parse_recipient_slug(slug: str) -> tuple[Any, str] | None:
+    """
+    Parses operands formatted in slug containing object ID or IDs.
+    Typical of "channel" or private message operands.
+
+    Doesn't parse the legacy stream slug.
+        e.g. "stream-name"
```

Comment:
We do need to support those for many likely use cases for this logic. But maybe that can be an optional parameter in the future.